File: projects/LeetCodeWP/src/main/java/com/hef/week09/sort/ShellSort.java
package com.hef.week09.sort;

import java.util.Arrays;

/**
 * @Date 2021/7/1
 * @Author lifei
 */
public class ShellSort {

    public static void main(String[] args) {
        Integer[] a = {9, 1, 4, 2, 8, 1, 3, 0, 2};
        System.out.println(Arrays.toString(a));
        ShellSort.sort(a);
        System.out.println(Arrays.toString(a));
    }

    public static void sort(Comparable[] a) {
        int k = 1, N = a.length;
        while (k<N/3) k = k*3 +1;
        while (k>=1) {
            for (int i=k; i<N; i++) {
                for (int j=i; j>=k && less(a, j, j-k); j-=k) {
                    exch(a, j, j-k);
                }
            }
            k = k/3;
        }
    }

    private static void exch(Comparable[] a, int i, int j) {
        Comparable t = a[i];
        a[i] = a[j];
        a[j] = t;
    }

    private static boolean less(Comparable[] a, int i, int j) {
        return a[i].compareTo(a[j])<0;
    }
}

File: document/js_wp/ShellSort.js

// 希尔排序，基于插入排序
function shellSort(a) {
    if (!a || a.length==0) {
        return;
    }
    let k = 1;
    while (k<Math.floor(a/3)) {
        k = k * 3 + 1;
    }

    while (k>=1) {
        for (let i=k; i<a.length; i++) {
            let j = i-k;
            let v = a[i];
            for (;j>=0; j-=k) {
                if (less(v, a[j])) {
                    a[j+k] = a[j];
                }else {
                    break
                }
            }
            a[j+k] = v;
        }
        k = Math.floor(k/3)
    }
}

function less(v1, v2) {
    if (v1<v2) {
        return true;
    }else {
        return false;
    }
}

// let a = [3, 6, 1, 2, 9, 0, 1, 2, 3]
let a = [3, 6, 1, 2, 9, 0, 1, 2, 3, 3, 3, 1, 2, 0, 0, 9, 3, 6, 1, 2, 9, 0, 1, 2, 3, 3, 3, 1, 2, 0, 0, 9]
console.log(a);
shellSort(a);
console.log(a);